### DISCUSSION TITLE
Automatic app access based on thirdPartyProfile

### DISCUSSION THREAD

**Discussion by Juan-A-Jimenez** [Budibase/budibase#7340] [Help]
Hi guys,

I am currently working with a Budibase setup together with a Keycloak ldp. Whenever we first login via SSO, a new budibase user is created. I've figured out that the user in the couch DB includes lots of extra information from out Keycloak within the attribute "thirdPartyProfile", like the external ID, the keycloak groups and permission and more. I've also figured out, that then I add a new custom column with the name "thirdPartyProfile" to the Budibase internal "Users" table, it shows me the information there and I can access it in the app.

Now to the questions: 
1. I want to assign access to certain apps based on the Keycloak roles in "thirdPartyProfile", I've tried using an automation, but I can't select the built-in "Users" table as a automation trigger.
I have tried connecting budibases internal DB via the "couchDB" integration as an "external" datasource, but I could not get that working. I can't get any query to return values from there. Have you guys connected the budibase couchdb as an "external" db before to access budibase objects directly?
I think I would need that to write the role/app assignment into the Users objects directly to grant them access to apps.

2. Have you guys managed to setup "real" SSO for budibase, in a way where when you are already logged in to another app on the same idp, you open budibase and don't have to enter your credentials again, but are automatically logged in? We can't get that to work on keycloak, we are missing an expert here.

3. I mentioned you can add a custom column to the "Users" table like "thirdPartyProfile" that will let you access the user json content of that key. Furthermore, you can also add columns even with sub-keys of that like "thirdPartyProfile.groups" which lets me access the groups directly in the users table, without needing to go through the whole "thirdPartyProfile" json object for that. However, here is the bug: When trying to access this table column with the name (key) "thirdPartyProfile.groups" through handlebar bindings, it's always empty, even though the Users table shows values there. My solution is to go though javascript like "return $("thirdPartyProfile").groups" which works fine but isn't optimal.

**Comment by melohagan** [comment]
Hey @Juan-A-Jimenez 

Some thoughts:

1. This is a custom RBAC setup and I don't think its possible currently.
2. When you attempt to log in via a configured SSO provider like google or azure, it will show your logged in account and prompt you to click the one you want.
3. What are you hoping to use the "thirdPartyProfile" for?

**Reply by PhilippFr** [reply]
The user objects in the internal budibase couchDB include an object with the key "thirdPartyProfile" that has lots of information from our idp keycloak, most importantly, the roles the user is assigned to in keycloak. If we could make an automation listen to the creation or update of a user. We can get these roles and set app permissions for the user via the REST API. This would be an effective RBAC.

The only issue with that is, we currently have no way of triggering an automatomation when an user is created or update. Hence our idea was to include the internal budibase couch DB, as an "external data connection" so we might be able to have an automation trigger on it.

We were not able to include the internal couchDB as an "external source" however. There unfortunately is very little documentation of the couchdb connector, and how to write queries for it.
